programs/mpl-core/src/plugins/utils.rs

```diff
@@ -415,6 +488,130 @@ pub fn delete_plugin<'a, T: DataBlob>(
+                        {
+                            Ok(address) => address,
+                            Err(_) => return Err(MplCoreError::DeserializationError.into()),
+                        }
+                }
+                ExternalPluginKey::DataStore(authority) => {
+                    let authority_offset = record
+                        .offset
+                        .checked_add(1)
+                        .ok_or(MplCoreError::NumericalOverflow)?;
+                    authority
+                        == &match Authority::deserialize(
+                            &mut &account.data.borrow()[authority_offset..],
+                        ) {
+                            Ok(authority) => authority,
+                            Err(_) => return Err(MplCoreError::DeserializationError.into()),
+                        }
+                }
+            })
+        {
+            result = Some(i);
+            break;
+        }
+    }
+
+    if let Some(index) = result {
+        let registry_record = plugin_registry.external_registry.remove(index);
+        let serialized_registry_record = registry_record.try_to_vec()?;
+
+        // Fetch the offset of the plugin to be removed.
+        let plugin_offset = registry_record.offset;
+        let plugin = ExternalPlugin::load(account, plugin_offset)?;
+        let serialized_plugin = plugin.try_to_vec()?;
+
+        // Get the offset of the plugin after the one being removed.
+        let next_plugin_offset = plugin_offset
+            .checked_add(serialized_plugin.len())
+            .ok_or(MplCoreError::NumericalOverflow)?;
+
+        // Calculate the new size of the account.
+        let new_size = account
+            .data_len()
+            .checked_sub(serialized_registry_record.len())
+            .ok_or(MplCoreError::NumericalOverflow)?
+            .checked_sub(serialized_plugin.len())
+            .ok_or(MplCoreError::NumericalOverflow)?;
+
+        let new_registry_offset = header
+            .plugin_registry_offset
+            .checked_sub(serialized_plugin.len())
+            .ok_or(MplCoreError::NumericalOverflow)?;
+
+        let data_to_move = header
+            .plugin_registry_offset
+            .checked_sub(next_plugin_offset)
+            .ok_or(MplCoreError::NumericalOverflow)?;
+
+        //TODO: This is memory intensive, we should use memmove instead probably.
+        let src = account.data.borrow()[next_plugin_offset..].to_vec();
+        sol_memcpy(
+            &mut account.data.borrow_mut()[plugin_offset..],
+            &src,
+            data_to_move,
+        );
+
+        header.plugin_registry_offset = new_registry_offset;
+        header.save(account, asset.get_size())?;
+
+        // Move offsets for existing registry records.
+        for record in &mut plugin_registry.external_registry {
+            if plugin_offset < record.offset {
+                record.offset -= serialized_plugin.len()
+            }
+        }
```

Comment:
I think what is worse here is we need to also move all the data_offset within the external plugin.  This is originally why I wanted to have the `data_offset` at the `ExternalRegistryRecord` level as well, but maybe its not so bad to do it per plugin especially if done in a helper.

Reply:
Getting fixed in https://github.com/metaplex-foundation/mpl-core/pull/87